servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/BaseHostSelector.java

```diff
@@ -67,21 +68,38 @@ protected final Single<C> noActiveHostsFailure(List<Host<ResolvedAddress, C>> us
                 this.getClass(), "selectConnection(...)"));
     }
 
+    // TODO: this could really be the core method on `Host` other than the nullable part. We could use Optional...
+    //  The API of passing in the status is weird but we could either live with it or just ask the question again
+    //  if we need to make a new connection.
+    protected final @Nullable Single<C> selectFromHost(
+            Host<ResolvedAddress, C> host, Host.Status status, Predicate<C> selector,
+            boolean forceNewConnectionAndReserve, @Nullable ContextMap contextMap) {
+        // First see if we can get an existing connection regardless of health status.
+        if (!forceNewConnectionAndReserve) {
+            C c = host.pickConnection(selector, contextMap);
+            if (c != null) {
+                return succeeded(c);
+            }
+        }
+        // We can only create a new connection if the host is active. It's possible for it to think that
+        // it's healthy based on having connections but not being active but we weren't able to pick an
+        // existing connection.
+        return status.active ? host.newConnection(selector, forceNewConnectionAndReserve, contextMap) : null;
+    }
+
     private Single<C> noHostsFailure() {
         return failed(Exceptions.StacklessNoAvailableHostException.newInstance(
                 "No hosts are available to connect for " + targetResource + ".",
                 this.getClass(), "selectConnection(...)"));
     }
 
-    private static <ResolvedAddress, C extends LoadBalancedConnection> boolean allUnhealthy(
+    private static <ResolvedAddress, C extends LoadBalancedConnection> boolean anyHealthy(
             final List<Host<ResolvedAddress, C>> usedHosts) {
-        boolean allUnhealthy = !usedHosts.isEmpty();
         for (Host<ResolvedAddress, C> host : usedHosts) {
-            if (!host.isUnhealthy()) {
-                allUnhealthy = false;
-                break;
+            if (host.status(false).healthy) {
```

Comment:
i feel that this is now harder to read, as a descriptive accessor. wdyt?
either use a descriptive variable ie. `WITHOUT_FORCED_CONNECTION` or consider improving the method name, ie. `statusWithForceConnection(...)`

Reply:
The name didn't bug me, although the parameter in general does. Since we're not sharing the status result with the `BaseSelector.selectFromHost` method anymore, I thin it's probably better to simple go back to two methods: `isHealthy()` and `isActive`.

Reply:
I went back to boolean methods as discussed, although I change the methods to have more descriptive names. Let me know if that doesn't work for you.